src/cli/crd/generate.test.ts

```diff
@@ -0,0 +1,202 @@
+    });
+  });
+
+  describe("extractDetails", () => {
+    const createProjectWithFile = (name: string, content: string) => {
+      const project = new Project();
+      return project.createSourceFile(name, content);
+    };
+
+    it("should extract plural, scope, and shortName from the details object", () => {
+      const file = createProjectWithFile(
+        "temp.ts",
+        `
+        const details = { plural: "widgets", scope: "Namespaced", shortName: "wd" };
+      `,
+      );
+
+      const details = extractDetails(file);
+      expect(details).toEqual({
+        plural: "widgets",
+        scope: "Namespaced",
+        shortName: "wd",
+      });
+    });
+
+    it("should throw if 'scope' is invalid", () => {
+      const file = createProjectWithFile(
+        "bad.ts",
+        `
+        const details = { plural: "widgets", scope: "BadScope", shortName: "wd" };
+      `,
+      );
+
+      expect(() => extractDetails(file)).toThrow(
+        `'scope' must be either "Cluster" or "Namespaced", got "BadScope"`,
+      );
+    });
+
+    it("should throw if details variable is missing", () => {
+      const file = createProjectWithFile("noDetails.ts", `const somethingElse = {};`);
+      expect(() => extractDetails(file)).toThrow(`Missing 'details' variable declaration.`);
+    });
+  });
+
+  describe("processSourceFile", () => {
+    const createProjectWithFile = (name: string, content: string) => {
+      const project = new Project();
+      return project.createSourceFile(name, content);
+    };
+
+    it("should skip file with no kind comment", () => {
+      const consoleWarn = jest.spyOn(console, "warn").mockImplementation(() => {});
+      const file = createProjectWithFile(
+        "test.ts",
+        `
+        const details = { plural: "widgets", scope: "Cluster", shortName: "wd" };
+        export interface SomethingSpec {}
+      `,
+      );
+
+      processSourceFile(file, "v1", "/output");
+      expect(consoleWarn).toHaveBeenCalledWith(
+        expect.stringContaining("missing '// Kind: <KindName>' comment"),
+      );
+    });
+
+    it("should skip file with missing Spec interface", () => {
+      const consoleWarn = jest.spyOn(console, "warn").mockImplementation(() => {});
+      const file = createProjectWithFile(
+        "test.ts",
+        `
+        // Kind: Something
+        const details = { plural: "widgets", scope: "Cluster", shortName: "wd" };
+      `,
+      );
+
+      processSourceFile(file, "v1", "/output");
+      expect(consoleWarn).toHaveBeenCalledWith(
+        expect.stringContaining("missing interface SomethingSpec"),
+      );
+    });
+
+    it("should generate a CRD YAML file for valid input", () => {
+      const writeFileSync = jest.spyOn(fs, "writeFileSync").mockImplementation(() => {});
+      const consoleLog = jest.spyOn(console, "log").mockImplementation(() => {});
+      const file = createProjectWithFile(
+        "valid.ts",
+        `
```

Comment:
Storing relatively simple files as string in a test is okay, but I wonder if we can store these as _actual_ files and read them in during a test. Maybe not required if we mock file read operations.

The thing I want to avoid is a large, bespoke test string like we've seen elsewhere in Pepr. One example of this anti-pattern is `rbac.test.ts` before we created `defaultTestObjects`. Thoughts?

Reply:
see #2067